plugin/plugins/wasmipld/wasmipld.go

```diff
@@ -0,0 +1,153 @@
+package wasmipld
+
+import (
+	"io/ioutil"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/multicodec"
+
+	mc "github.com/multiformats/go-multicodec"
+
+	"github.com/ipfs/go-ipfs/plugin"
+	"github.com/mitchellh/mapstructure"
+
+	wasmbind "github.com/aschmahmann/wasm-ipld/gobind"
+)
+
+// Plugins is exported list of plugins that will be loaded
+var Plugins = []plugin.Plugin{
+	&wasmipld{},
+}
+
+type wasmipld struct{}
+
+var _ plugin.PluginIPLD = (*wasmipld)(nil)
+var _ plugin.PluginIPLDADL = (*wasmipld)(nil)
+
+func (*wasmipld) Name() string {
+	return "ipld-wasmipld"
+}
+
+func (*wasmipld) Version() string {
+	return "0.0.1"
+}
+
+type WasmIPLDConfig struct {
+	Codecs []WasmIPLDCodecConfig
+	ADLs   []WasmIPLDADLConfig
+}
+
+type WasmIPLDCodecConfig struct {
+	Code     string
+	Encode   bool
+	Decode   bool
+	WasmPath string
+}
+
+type WasmIPLDADLConfig struct {
+	Name     string
+	WasmPath string
+}
+
+func (*wasmipld) Init(env *plugin.Environment) error {
+	config := env.Config
+	if config == nil {
+		return nil
+	}
+
+	var cfg WasmIPLDConfig
+	// Note: This dependency is just for convenience and is in our go.sum anyway
+	// It can go away if we upstream this into the main config instead of keeping it as a plugin
+	if err := mapstructure.Decode(config, &cfg); err != nil {
+		return err
+	}
+
+	registry = &wasmRegistry{}
+
+	for _, c := range cfg.Codecs {
+		wasm, err := ioutil.ReadFile(c.WasmPath)
+		if err != nil {
+			return err
+		}
+
+		var code mc.Code
+		if err := code.Set(c.Code); err != nil {
+			return err
+		}
+
+		codecRegistration := codecReg{
+			code:   code,
+			wasm:   wasm,
+			encode: c.Encode,
+			decode: c.Decode,
+		}
+		registry.Codecs = append(registry.Codecs, codecRegistration)
+	}
+
+	for _, a := range cfg.ADLs {
+		wasm, err := ioutil.ReadFile(a.WasmPath)
+		if err != nil {
+			return err
+		}
+
+		adlRegistration := adlReg{
+			name: a.Name,
+			wasm: wasm,
+		}
+
+		registry.ADLs = append(registry.ADLs, adlRegistration)
+	}
+
+	return nil
+}
+
+const fuelPerOp = 10_000_000
```

Comment:
maybe make configurable?

Reply:
ya, makes sense. Just left it there for now, not sure what makes sense for limits. Could certainly allow limits to be per-module or something if people want to mess with things. Could also remove the limits by default 🤷 

None of our codecs or ADLs have any sort of limits like this at the moment. It's worth considering what this means in terms of helping users make their code predictably deployable. Wouldn't want people to be surprised that some of their data loaded with implementation A, but not B due to limits here. Obviously people can do whatever they want as with block size limits, but we probably want some reasonably large safe boundaries to let people work with.